### ISSUE TITLE
Make mfsdp optimizer primitive self-contained (follow-up to PR #26): 薄热核自研 + 复用 fsdp2 util,零 megatron.core import

### ISSUE THREAD

**Issue by ISEEKYAN** [ISEEKYAN/Megatron-LM#87]
# Make the `mfsdp` optimizer primitive self-contained (follow-up to PR #26)

## 背景 / Background
PR #26 给 MLite 加了 Megatron-FSDP (`mfsdp`) optimizer primitive(`experimental/lite/megatron/lite/primitive/optimizers/mfsdp/`),已有 CPU unit(`test_mfsdp.py` 5 passed)+ 2-GPU FSDP2 parity smoke(job `12728610`,loss/grad_norm/max_param_diff 逐位对齐)。
**问题**:当前 `mfsdp/` 直接 import 了 **14 处 `megatron.core` 符号**,不 self-contained。

## 目标 / Goal
把 `mfsdp` 做成 **self-contained**:`mfsdp/` 内**零 `megatron.core` import**(fused op 走独立封装),同时**保持逐位 parity 与性能**。

## 硬约束 / Constraints
- ✅ `mfsdp/` 保持**独立文件夹 + 真 M-FSDP 算法/训练路径**(不折进 fsdp2、不复用 fsdp2 的 `fully_shard` 训练路径)。
- 🚫 **`fsdp2/` 原有代码/路径一行不改**。
- ✅ **允许 `mfsdp/` import `fsdp2/` 里已有的通用工具函数**(DTensor helper / state-dict util 等),不重复造轮子。
- ✅ 真正的 fused op 走独立 `fused_ops.py` 封装(TE→Apex→Torch fallback 顺序)。

## Import 审计（要处理的 mcore 依赖）
| 文件 | 当前 mcore 依赖 |
|---|---|
| `backend.py` | `preprocess_state_dict_for_uneven_dtensor`(uneven-DTensor checkpoint canonicalization) |
| `grad_norm.py` | `param_is_not_tensor_parallel_duplicate` / `param_is_not_shared` / `to_local_if_dtensor` |
| `optimizer.py` | `DistributedDataParallelConfig` / `FullyShardedDataParallel` / `finalize_model_grads` / `get_megatron_optimizer` 等 |
| `patches.py` | `param_and_grad_buffer` / `megatron_fsdp` 模块(AllGather / `start_param_sync` patch) |
| `process_groups.py` | `mcore_fsdp_adapter` / `ProcessGroupCollection` |
| 其余 6 文件 | 无直接 mcore import |
| **隐藏旁路** | `optimizer.py → megatron_wrap.py` 会**间接**再拉 mcore parallel state / OptimizerConfig / TransformerConfig / PG collection —— 也必须切断 |

## 推荐方案 / Recommended approach
**受限版方案 B:薄 M-FSDP 热核(保真自研)+ 复用 fsdp2 通用原语。**

**关键洞见(降低了这件事的难度)**:
1. **`DistributedOptimizer.__init__` 在 `use_megatron_fsdp=True` 时提前 return** → 大部分通用 distributed optimizer **不进 M-FSDP 热路径**。
2. **checkpoint(uneven-DTensor canonicalization)和 process groups 都不在训练热循环** → 安全**复用 fsdp2 util / 抽成共享 checkpoint primitive**(uneven-DTensor 逻辑别删,抽成共享 primitive)。
3. **真正的训练热路径很窄**:`patches.py` 的 **AllGatherPipeline / `start_param_sync` / `param_and_grad_buffer` overlap+bucketing** + 少量 grad reduce 调度 —— **只有这块必须保真自研并实测性能**。

**逐处处理原则**:
- **热路径(patches overlap / sharding strategy)** → 保真自研,留在 `mfsdp/`。
- **非热路径(checkpoint / PG / grad_norm 判定 / config)** → 复用 `fsdp2/` util 或 torch 等价 / 抽成共享 primitive。
- **`megatron_wrap` 间接旁路** → 换本地 config dataclass + 本地 PG 构造,切断。

**方案 A(全保真移植·最小传递闭包)作为回退**:若受限版 B 的自研热核 parity 或性能不达标,退回"以 PR#26 为 oracle 把最小闭包移进 `mfsdp/`"。

## 优劣对比 / Comparison
| 维度 | 方案A(全移植) | 方案B(薄热核+复用fsdp2) |
|---|---:|---:|
| 工作量/新增代码少 | 1/5 | **4/5** |
| 维护成本+升级脱钩 | 2/5 | **5/5** |
| 性能匹配确定性 | **4/5** | 3/5 |
| self-contained 纯度 | **5/5** | 4/5 |
| 逐位 parity 验证易 | **4/5** | 2/5 |
| 可测试性 | 3/5 | **5/5** |

## 验证计划 / Validation
1. **AST/poison-import guard**(单测):禁止 `mfsdp/`(除 `fused_ops.py`)出现任何 `megatron.core` 显式/动态/经 `megatron_wrap` 间接 import;`import mfsdp` 后 `sys.modules` 不得含 `megatron.core`。
2. **CPU unit**:`test_mfsdp.py` 保持 `5 passed`、退出码 0;新增本地 helper 测试通过。
3. **2-GPU bitwise parity**(GPU,唯一 GPU 步):同 job `12728610` 的容器/依赖栈,1 节点 2 GPU,跑 FSDP2 parity smoke,**收紧到逐位相等**:loss 完全相等、grad_norm 完全相等、所有完整参数 `torch.equal`、`max_param_abs_diff == 0`;证据含真实 job id + `sacct` COMPLETED/0:0 + pytest `1 passed` 非 skip。

## 验收 / Acceptance
- `mfsdp/` 内 `megatron.core` import = 0(AST 守卫红→绿);`fsdp2/` 零改动;`mfsdp` 保持独立文件夹与真 M-FSDP 路径。
- CPU unit + 2-GPU 逐位 parity 全过(与 PR#26 job 12728610 一致)。

> 参考:PR #26 https://github.com/ISEEKYAN/Megatron-LM/pull/26

**Comment by ISEEKYAN** [comment]
## 📋 同事版实施指南：14 处 mcore 依赖逐处怎么改（说人话）

> 核心思路一句话:**mfsdp 保持自己的文件夹和真 M-FSDP 算法;把"不在训练热循环"的部分复用 fsdp2 现成工具/抽成共享件,只有真正的性能热路径保留自研。fsdp2 自己的代码一行不改。**

先分清哪些是"热路径"(每步训练都跑、决定速度)vs"管道"(建立/存取,不进热循环):

### 🟢 管道类(不影响性能,复用/换掉,别自己重造)

**1. `process_groups.py` — `mcore_fsdp_adapter` / `ProcessGroupCollection`**
- 是什么:建进程组(哪些卡组成一个通信组)。只在启动时跑一次,不进训练循环。
- 怎么改:建一个本地的 PG dataclass,用 `torch.distributed` 的 DeviceMesh 直接建;或直接 import fsdp2 里已有的 PG 工具。零性能影响。

**2. `backend.py` — `preprocess_state_dict_for_uneven_dtensor`**
- 是什么:存/读 checkpoint 时,把不均匀分片的 DTensor 规整成能落盘的格式。只在 save/load 时跑,不进训练循环。
- 怎么改:**别删**(它是必需的正确性逻辑)。把这段抽成一个**共享 checkpoint 工具**(放一个两边都能 import 的位置),`mfsdp` 和 `fsdp2` 都从那里调。这样不重复、也不 import mcore。

**3. `grad_norm.py` — `param_is_not_tensor_parallel_duplicate` / `param_is_not_shared` / `to_local_if_dtensor`**
- 是什么:算梯度范数前,判断一个参数要不要计入(TP 重复的/shared 的不能重复算)+ 把 DTensor 转成本地张量。都是小判断,一次 all-reduce。
- 怎么改:用参数自己带的属性(`param.tensor_model_parallel`、shared 标记)本地判断,几行就够;`to_local_if_dtensor` 用 torch 的 `DTensor.to_local()`。或直接复用 fsdp2 里已有的等价工具。

**4. `optimizer.py` 的非热路径部分 — `DistributedDataParallelConfig` / `get_megatron_optimizer` 等**
- **关键发现**:mcore 的 `DistributedOptimizer.__init__` 在 mfsdp 模式下**提前 return**,大部分通用 optimizer 逻辑**根本不进 mfsdp 热路径**。
- 怎么改:config 换成本地小 dataclass;那些只在构建期用的 optimizer 组装逻辑,用 fsdp2 现成的或本地等价重写。

**5. 隐藏旁路 — `optimizer.py → megatron_wrap.py`**
- 是什么:optimizer.py 经 `megatron_wrap` 又间接把 mcore 的 parallel state / OptimizerConfig / TransformerConfig / PG collection 拉回来。
- 怎么改:必须切断——换成本地 config 构造 + 本地 PG,不再经 megatron_wrap 拉 mcore。

### 🔴 热路径类(性能来源,必须保真自研,不能糊弄)

**6. `patches.py` — `param_and_grad_buffer` / `AllGatherPipeline` / `start_param_sync`**
- 是什么:**这就是 M-FSDP 性能优势的核心**——参数 all-gather 和 param-sync 与计算的**重叠(overlap)+ 分桶(bucketing)**,每一步训练都跑。
- 怎么改:**保留自研**,把这套 buffer + AllGather 流水线 + param-sync 的逻辑**在 mfsdp 内本地实现**(不 import mcore 的 param_and_grad_buffer,自己写等价的)。**这块是重点工作量,也是性能成败关键,不能用通用实现替代。**

### 实施顺序（小步 TDD，每步可验）
1. **先加红测**:一个 AST 检查测试,扫 `mfsdp/`(除 fused_ops.py)里任何 `megatron.core` import → 现在红,做完变绿。
2. **先做管道(低风险)**:按 1→5 顺序,每改一处跑 `pytest experimental/lite/tests/unit/primitive/test_mfsdp.py`(保持 5 passed)。
3. **再做热核(高风险)**:第 6 处 patches,改完先 CPU 单测,再上 2-GPU parity。
4. **最终 GPU 逐位 parity**:同 PR#26 的 job `12728610` 容器,2 卡跑 `test_mfsdp_parity_smoke.py`,要求 loss/grad_norm/所有参数 `torch.equal`、`max_param_abs_diff == 0`。
5. **性能门**:热核改完实测吞吐,不达标就退回"把 mcore 的对应实现原样搬进来"(方案A 兜底),**不许放宽阈值蒙混**。

### 验证命令
```bash
# 静态边界(红→绿)
PYTHONPATH=experimental/lite pytest -q experimental/lite/tests/unit/primitive/test_mfsdp.py
# 确认 import 后无 mcore
python -c "import megatron.lite.primitive.optimizers.mfsdp as m; import sys; assert not any(k.startswith('megatron.core') for k in sys.modules)"
# 2-GPU 逐位 parity(Slurm,同 job 12728610 容器)
sbatch <2-gpu parity smoke>   # 要求 sacct COMPLETED 0:0 + pytest 1 passed 非 skip
```
